File: src/main/java/ru/megains/game/WorldRenderer.java
//package ru.megains.game;
//
//
//import ru.megains.engine.graph.Material;
//import ru.megains.engine.graph.RenderChunk;
//import ru.megains.engine.graph.Texture;
//import ru.megains.renderer.texture.TextureLoader;
//import ru.megains.game.blockdata.BlockPos;
//import ru.megains.game.entity.Entity;
//import ru.megains.game.world.World;
//import ru.megains.game.world.chunk.Chunk;
//import ru.megains.game.world.chunk.ChunkVoid;
//
//import java.util.HashMap;
//import java.util.Map;
//
//public class WorldRenderer {
//
//    public World world;
//
//    public HashMap<Long,RenderChunk> renderChunks;
//
//    public WorldRenderer(World world) {
//        this.world = world;
//
//
//    }
//
//    public HashMap<Long, RenderChunk> getRenderChunks() {
//        return renderChunks;
//    }
//
//    public void init(){
//
//
//
//        RenderChunk.material = new Material(new Texture("/textures/grassblock.png"));
//        RenderChunk.material.texture.setId(TextureLoader.blockTexture());
//        RenderChunk.material.ambient.set(0.5f, 0.5f, 0.5f, 1.0f);
//        RenderChunk.material.diffuse.set(0.8f, 0.8f, 0.8f, 1.0f);
//        RenderChunk.material.specular.set(0.1f, 0.1f, 0.1f, 1.0f);
//        RenderChunk.material.emission.set(0f, 0f, 0f, 1.0f);
//        RenderChunk.material.shininess = 20.0f;
//
//        renderChunks = new HashMap<>();
//
//
//        for(Map.Entry<Object,Chunk> i: world.chunks().seq()){
//           renderChunks.put((Long) i.getKey(),new RenderChunk(i.getValue()));
//        }
//
//
//    }
//
//
//
//
//
//    public void cleanUp(){
//        renderChunks.values().forEach(RenderChunk::cleanUp);
//    }
//
//    public void reRender(BlockPos pos) {
//
//
//        int x = pos.x()>>4;
//        int y = pos.y()>>4;
//        int z = pos.z()>>4;
//        getRenderChunkRen(x,y,z).reRender();
//        getRenderChunkRen(x+1,y,z).reRender();
//        getRenderChunkRen(x-1,y,z).reRender();
//        getRenderChunkRen(x,y+1,z).reRender();
//        getRenderChunkRen(x,y-1,z).reRender();
//        getRenderChunkRen(x,y,z+1).reRender();
//        getRenderChunkRen(x,y,z-1).reRender();
//
//    }
//
//    public RenderChunk getRenderChunkRen(int x, int y, int z){
//
//        long i =(x & 16777215l) << 40 | (z & 16777215l) << 16 | (y & 65535L);
//        return renderChunks.containsKey(i)?renderChunks.get(i):voidRender;
//
//    }
//    public static RenderChunk voidRender = new RenderChunk(new ChunkVoid());
//
//
//
//    public void renderEntity(Entity entity) {
//
////        if(entity.mesh==null){
////            createEntityMesh(entity);
////        }
////
////        entity.mesh.render();
//
//    }
//    public void createEntityMesh(Entity entity){
//
////        AxisAlignedBB AABB = entity.cube;
////        float minX = AABB.getMinX();
////        float minY = AABB.getMinY();
////        float minZ = AABB.getMinZ();
////        float maxX = AABB.getMaxX();
////        float maxY = AABB.getMaxY();
////        float maxZ = AABB.getMaxZ();
////
////
////
////
////        Material material = new Material(new Texture(TextureLoader.blockTexture()) );
////
////        MeshMaker$ mm =MeshMaker$.MODULE$;
////        mm.startMakeTriangles();
////        mm.setCurrentIndex();
////
////        mm.addVertexWithUV(minX, minY, minZ, 0, 1);
////        mm.addVertexWithUV(minX, maxY, minZ, 0, 0);
////        mm.addVertexWithUV(maxX, maxY, minZ, 1, 0);
////        mm.addVertexWithUV(maxX, minY, minZ, 1, 1);
////        mm.addIndex(0);
////        mm.addIndex( 1 );
////        mm.addIndex( 2);
////        mm.addIndex(0);
////        mm.addIndex(2);
////        mm.addIndex(3);
////
////        entity.mesh = mm.makeMesh();
////        entity.mesh.setMaterial(material);
//
//
//    }
//
//    public void update() {
//        for(RenderChunk chunk:renderChunks.values()){
//            chunk.update();
//
//        }
//    }
//}

File: src/main/java/ru/megains/engine/graph/WorldRenderer.scala
package ru.megains.engine.graph

import ru.megains.engine.Frustum
import ru.megains.game.blockdata.{BlockWorldPos, MultiBlockPos}
import ru.megains.game.entity.item.EntityItem
import ru.megains.game.register.GameRegister
import ru.megains.game.util.BlockAndPos
import ru.megains.game.world.World
import ru.megains.game.world.chunk.{Chunk, ChunkVoid}
import ru.megains.managers.TextureManager
import ru.megains.renderer.mesh.{Mesh, MeshMaker}

import scala.collection.mutable


class WorldRenderer(val world: World, val textureManager: TextureManager) {
    world.worldRenderer = this

    val renderChunks: mutable.HashMap[Long, RenderChunk] = new mutable.HashMap[Long, RenderChunk]
    var voidRender: RenderChunk = new RenderChunk(new ChunkVoid, textureManager)
    var mesh: Mesh = _

    def init() {
        world.chunks.foreach((chunk: (Long, Chunk)) => renderChunks.put(chunk._1, new RenderChunk(chunk._2, textureManager)))
    }

    def reRender(pos: BlockWorldPos) {
        val x: Int = pos.worldX >> 4
        val y: Int = pos.worldY >> 4
        val z: Int = pos.worldZ >> 4
        getRenderChunkRen(x, y, z).reRender()
        getRenderChunkRen(x + 1, y, z).reRender()
        getRenderChunkRen(x - 1, y, z).reRender()
        getRenderChunkRen(x, y + 1, z).reRender()
        getRenderChunkRen(x, y - 1, z).reRender()
        getRenderChunkRen(x, y, z + 1).reRender()
        getRenderChunkRen(x, y, z - 1).reRender()
    }

    def getRenderChunkRen(x: Int, y: Int, z: Int): RenderChunk = {
        val i: Long = (x & 16777215l) << 40 | (z & 16777215l) << 16 | (y & 65535L)

        if (renderChunks.contains(i)) renderChunks(i) else voidRender
    }

    def cleanUp(): Unit = renderChunks.values.foreach(_.cleanUp)

    def updateBlockBounds(bp: BlockAndPos): Unit = {
        if (mesh != null) {
            mesh.cleanUp()
            mesh = null
        }
        val mm = MeshMaker.getMeshMaker
        val aabb = bp.block.getSelectedBoundingBox(bp.pos, MultiBlockPos.default).expand(0.01f, 0.01f, 0.01f)

        val minX = aabb.getMinX
        val minY = aabb.getMinY
        val minZ = aabb.getMinZ
        val maxX = aabb.getMaxX
        val maxY = aabb.getMaxY
        val maxZ = aabb.getMaxZ



        mm.startMake(1)
        mm.addVertex(minX, minY, minZ)
        mm.addVertex(minX, minY, maxZ)
        mm.addVertex(minX, maxY, minZ)
        mm.addVertex(minX, maxY, maxZ)
        mm.addVertex(maxX, minY, minZ)
        mm.addVertex(maxX, minY, maxZ)
        mm.addVertex(maxX, maxY, minZ)
        mm.addVertex(maxX, maxY, maxZ)

        mm.addIndex(0, 1)
        mm.addIndex(0, 2)
        mm.addIndex(0, 4)

        mm.addIndex(6, 2)
        mm.addIndex(6, 4)
        mm.addIndex(6, 7)

        mm.addIndex(3, 1)
        mm.addIndex(3, 2)
        mm.addIndex(3, 7)

        mm.addIndex(5, 1)
        mm.addIndex(5, 4)
        mm.addIndex(5, 7)


        mesh = mm.makeMesh()


    }

    def renderBlockBounds(shaderProgram: ShaderProgram): Unit = if (mesh != null) mesh.render(shaderProgram, textureManager)

    def renderEntitiesItem(frustum: Frustum, transformation: Transformation, sceneShaderProgram: ShaderProgram): Unit = {


        world.entities.filter(_.isInstanceOf[EntityItem]).foreach(
            (entity) => {
                if (frustum.cubeInFrustum(entity.body)) {
                    val modelViewMatrix = transformation.buildEntityModelViewMatrix(entity)
                    sceneShaderProgram.setUniform("modelViewMatrix", modelViewMatrix)
                    GameRegister.getItemRender(entity.asInstanceOf[EntityItem].item).renderInWorld(sceneShaderProgram, textureManager)
                }
            }
        )
    }

}
